### PR TITLE
Port AddAffiliateLocationExtensions.

### PR DESCRIPTION
**Pull request by nwbirnie**
See https://github.com/googleads/google-ads-dotnet/blob/master/examples/Extensions/AddAffiliateLocationExtensions.cs

Change-Id: I436778b93d4648737c1b1ce52cfae7a918314c33

### CODE REVIEW

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+        names = ArgumentNames.CHAIN_ID,
+        description =
+            "The retail chain ID. See"
+                + " https://developers.google.com/adwords/api/docs/appendix/codes-formats#chain-idsfor"
+                + " a complete list of valid retail chain IDs",
+        required = true)
+    private long chainId;
+  }
+
+  private static final int MAX_FEEDMAPPING_RETRIEVAL_ATTEMPTS = 10;
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:81]: Move const to top of class

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+
+  public static void main(String[] args) throws IOException {
+    AddAffiliateLocationExtensionsParams params = new AddAffiliateLocationExtensionsParams();
+    if (!params.parseArguments(args)) {
+
+      // Either pass the required parameters for this example on the command line, or insert them
+      // into the code here. See the parameter class definition above for descriptions.
+      params.customerId = Long.parseLong("INSERT_CUSTOMER_ID_HERE");
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:89]: Missing other params

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+      }
+    }
+  }
+
+  /** Runs the example. */
+  private void runExample(
+      GoogleAdsClient googleAdsClient, long customerId, long campaignId, long chainId) {
+    String feedResourceName =
+        createAffiliateLocationExtensionFeed(googleAdsClient, customerId, chainId);
+    // After the completion of the feed creation operation above the added feed will not
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:128]: Suggested edit (assuming I'm understand the flow correctly here): The feed created above will not be available for use in a campaign feed until the associated feed mappings have been created by Google. 
Attempts to retrieve the feedMapping using an exponential back-off policy until the feed mappings are created.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+  /** Creates an affiliate location extension feed. */
+  private String createAffiliateLocationExtensionFeed(
+      GoogleAdsClient googleAdsClient, long customerId, long chainId) {
+    // Deletes all existing location extension feeds. This is an optional step,
+    // but is required for this code example to run correctly more than once. This is because:
+    //   1. Google Ads only allows one location extension feed per email address.
+    //   2. A Google Ads account cannot have a location extension feed and an affiliate
+    // location extension feed at the same time.
+    deleteLocationExtensionFeeds(googleAdsClient, customerId);
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:145]: Suggest replacing delete with remove globally.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+                    .setRelationshipType(AffiliateLocationFeedRelationshipType.GENERAL_RETAILER))
+            // Since this feed's contents will be managed by Google,
+            // you must set its origin to GOOGLE.
+            .setOrigin(FeedOrigin.GOOGLE)
+            .build();
+
+    // Constructs an operation to create the feed.
+    FeedOperation op = FeedOperation.newBuilder().setCreate(feed).build();
+
+    // Connects to the API.
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:165]: Nit: change to "Creates the FeedServiceClient" (consistent with other examples).

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+      // Displays the results.
+      String resourceName = response.getResults(0).getResourceName();
+      System.out.printf(
+          "Affliate location extension feed created with resource name: %s.%n", resourceName);
+
+      return resourceName;
+    }
+  }
+
+  /** Deletes a location extension feed. */
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:181]: Deletes all? (same for comments below)

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+  private void deleteLocationExtensionFeeds(GoogleAdsClient googleAdsClient, long customerId) {
+    // Deletes a location extension feed. Does this with the following steps:
+    //   1. Delete the CustomerFeed so that the location extensions from the feed stop serving.
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:184]: Include a step for retrieving feeds?

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+    try (FeedServiceClient feedServiceClient =
+        googleAdsClient.getLatestVersion().createFeedServiceClient()) {
+      feedServiceClient.mutateFeeds(String.valueOf(customerId), removeOperations);
+    }
+  }
+
+  /** Retrieves all location extension feeds from a specified customer ID. */
+  private List<Feed> getLocationExtensionFeeds(GoogleAdsClient googleAdsClient, long customerId) {
+    String query =
+        "SELECT "
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:273]: nit: remove extra space after SELECT as there is a space before the first field

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+            + "  customer_feed.resource_name, "
+            + "  customer_feed.feed, "
+            + "  customer_feed.status, "
+            + "  customer_feed.matching_function.function_string "
+            + "FROM "
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:278]: Nit: remove extra space after FROM as there is a space before customer_feed. You can also just make this one line.

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+            + "  customer_feed "
+            + "WHERE "
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:280]: Nit: remove extra space after WHERE

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+        feedServiceClient.mutateFeeds(String.valueOf(customerId), operations);
+      }
+    }
+  }
+
+  /** Retrieves the location extension feeds linked to a customer ID. */
+  private List<CustomerFeed> getLocationExtensionCustomerFeeds(
+      GoogleAdsClient googleAdsClient, long customerId) {
+    String query =
+        "SELECT "
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:313]: Same comments as above

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+        GrpcStatusCode.of(Code.DEADLINE_EXCEEDED),
+        true);
+  }
+
+  /** Retrieves the {@link FeedMapping} for a given feed. */
+  private Optional<FeedMapping> getFeedMapping(
+      GoogleAdsClient googleAdsServiceClient, long customerId, String feedResourceName) {
+    String query =
+        String.format(
+            "SELECT "
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:376]: Same comments as above

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+    // Waits for Google's servers to setup the feed with feed attributes and attribute mapping.
+    // This process is asynchronous, so we wait until the feed mapping is created, with exponential
+    // backoff.
+    while (numAttempts < MAX_FEEDMAPPING_RETRIEVAL_ATTEMPTS) {
+      Optional<FeedMapping> feedMapping =
+          getFeedMapping(googleAdsServiceClient, customerId, feedResourceName);
+
+      if (feedMapping.isPresent()) {
+        System.out.printf("Feed %s is now ready.%n", feedResourceName);
+        return feedMapping.get();
```
Comment (devchas) [inline google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:348]: Does this provide the resource name?

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java (hunk 3 above)
Comment (nwbirnie) [reply google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java]: Updated PTAL

File: google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java (hunk 13 above)
Comment (nwbirnie) [reply google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java:349]: No this provides the FeedMapping.

**Review by devchas — APPROVED** [review]

**Review by AnashOommen — APPROVED** [review]

**Pull request merged**